x-pack/plugins/cases/public/common/hooks.ts

```diff
@@ -5,11 +5,117 @@
  * 2.0.
  */
 
+import { useCallback, useEffect, useState } from 'react';
+
+import { useLocation, useHistory } from 'react-router-dom';
+
+import useLocalStorage from 'react-use/lib/useLocalStorage';
+import { parse, stringify } from 'query-string';
 import { STACK_APP_ID } from '../../common/constants';
 import { useCasesContext } from '../components/cases_context/use_cases_context';
 
+import { DEFAULT_QUERY_PARAMS } from '../containers/use_get_cases';
+
+import type {
+  LocalStorageQueryParams,
+  QueryParams,
+  ParsedUrlQueryParams,
+  UrlQueryParams,
+} from '../../common/ui/types';
+
+const parseUrlQueryParams = (search: string): UrlQueryParams => {
+  const parsedUrlParams: ParsedUrlQueryParams = parse(search);
+
+  const urlParams: UrlQueryParams = {
+    ...(parsedUrlParams.sortField && { sortField: parsedUrlParams.sortField }),
+    ...(parsedUrlParams.sortOrder && { sortOrder: parsedUrlParams.sortOrder }),
+  };
+
+  const intPage = parsedUrlParams.page && parseInt(parsedUrlParams.page, 10);
+  const intPerPage = parsedUrlParams.perPage && parseInt(parsedUrlParams.perPage, 10);
+
+  if (intPage) {
+    urlParams.page = intPage;
+  }
+
+  if (intPerPage) {
+    urlParams.perPage = intPerPage;
+  }
+
+  return urlParams;
+};
+
 export const useIsMainApplication = () => {
   const { appId } = useCasesContext();
 
   return appId === STACK_APP_ID;
 };
+
+export function useUrlState() {
+  const location = useLocation();
+  const history = useHistory();
+
+  const [queryParams, setQueryParams] = useState<UrlQueryParams>({});
+
+  const [localStorageQueryParams, setLocalStorageQueryParams] =
+    useLocalStorage<LocalStorageQueryParams>('cases.list.preferences');
+
+  const setUrlQueryParams = useCallback(
```

Comment:
I think we need to be careful about setting parameters in the URL when using the case list as a modal inside another page / app as this can have unwanted side effects. For example, when attaching ML embeddables to a case, the page is expecting job ID parameters to be present in the URL and the job selector flyout is incorrectly being opened when the URL is being manipulated.

<img width="1499" alt="image" src="https://user-images.githubusercontent.com/7405507/199302776-b7e2902b-98c8-4865-8473-d34c11ee9508.png">

It might make more sense to only persist page settings in the URL if you are on the Cases list page, and for the modal to always default to 10 cases per page for example.

Reply:
> I think we need to be careful about setting parameters in the URL when using the case list as a modal inside another page / app as this can have unwanted side effects. 

The current behavior should remove the modal-specific URL parameters as soon as the modal is unmounted.

When I noticed that the alerts page had some URL parameters of its own, I made it so that those would remain on the page and `page + perPage + sortField + sortOrder` disappear when you close the modal.

Did it behave differently with these job ID params you mention?

---

I did find a somewhat related bug though.

If we filter by status in the all-cases view and pick `Closed`, the `sortField` becomes `closedAt`.

This works fine in this view but if then we navigate to Alerts and pick `Add existing case to alert` we shouldn't be able to filter by that status there so the table doesn't work properly. Existing Cases imply `Open` or `In Progress` only.